server/market/market.go

```diff
@@ -116,35 +121,140 @@ func NewMarket(mktInfo *dex.MarketInfo, storage db.DEXArchivist, swapper Swapper
 		return nil, err
 	}
 
-	// Lock coins backing active orders.
-	baseCoins, quoteCoins, err := storage.ActiveOrderCoins(mktInfo.Base, mktInfo.Quote)
+	// Load existing book orders from the DB.
+	base, quote := mktInfo.Base, mktInfo.Quote
+	bookOrders, err := storage.BookOrders(base, quote)
 	if err != nil {
 		return nil, err
 	}
-	// TODO: first check that the coins aren't already locked by e.g. another market.
-	log.Debugf("Locking %d base asset coins.", len(baseCoins))
-	for oid, coins := range baseCoins {
-		log.Tracef(" - order %v: %v", oid, coins)
+
+	log.Infof("Loaded %d stored book orders.", len(bookOrders))
+	// Put the book orders in a map so orders that no longer have funding coins
+	// can be removed easily.
+	bookOrdersByID := make(map[order.OrderID]*order.LimitOrder, len(bookOrders))
+	for _, ord := range bookOrders {
+		bookOrdersByID[ord.ID()] = ord
+	}
+
+	baseCoins := make(map[order.OrderID][]order.CoinID)
+	quoteCoins := make(map[order.OrderID][]order.CoinID)
+
+ordersLoop:
+	for id, lo := range bookOrdersByID {
+		if lo.FillAmt > 0 {
+			// Order already matched with another trade, so it is expected that
+			// the funding coins are spent in a swap.
+
+			// In general, our position is that the server is not ultimately
+			// responsible for verifying that all orders have locked coins since
+			// the client will be penalized if they cannot complete the swap.
+			// The least the server can do is ensure funding coins for NEW
+			// orders are unspent and owned by the user, but there are other
+			// possibilities:
+
+			// TODO 1: We may consider following the chain of dex contracts and
+			// redeems to ensure that lo.Remaining() in value is unspent, or at
+			// least check that the initial coins were spent in a dex monitored
+			// tx with TxMonitored(..., spendingTx).
+
+			// TODO 2: The swapper could notify market when a redeem output is
+			// created so it can lock those coins if the order is still on the
+			// book (e.g. signal{orderID, redeemOutCoins} -> market). Otherwise
+			// the user could place new orders with these coins even though they
+			// are implicitly funding the still booked order with a remaining
+			// amount.
```

Comment:
I think we should look at how client handles locking wallet coins from swap chain to fund partially-filled orders that remain on the books.  @buck54321 maybe you know off hand.  I'd have to dig into it.

Reply:
The wallet's lock the coins in the `Fund` and `FundingCoins` methods. That includes unfilled orders on startup.

Reply:
> The current pattern is to put all utxos in the first match and then chain the change output, so if an order is partially filled, there will be not attempt to lock

No attempt to lock because the change output is in a different account or because the change outputs are locked immediately.  I'm just wondering about the case where an order is partially filled and goes back on the books (not chained right then), the swap contract creating change, then the user places a new order that it tries to fund with that swap contract change that was not consumed by a chained in another swap since the first order is still on the books.